modules/core/src/main/scala/ch/epfl/scala/debugadapter/internal/evaluator/RuntimeEvaluator.scala

```diff
@@ -0,0 +1,475 @@
+    /* -------------------------------------------------------------------------- */
+    @inline private def invokeStaticMethod(tree: StaticMethodTree): Safe[JdiValue] =
+      for {
+        args <- tree.args.map(evaluate).traverse
+        argsBoxedIfNeeded <- JdiObject.boxUnboxOnNeed(tree.method.argumentTypes(), args, frame)
+        result <- JdiClass(tree.on, frame.thread).invokeStatic(tree.method, argsBoxedIfNeeded)
+      } yield result
+
+    @inline private def invokePrimitiveMethod(
+        tree: PrimitiveMethodTree
+    ): Safe[JdiValue] =
+      for {
+        lhs <- evaluate(tree.lhs).flatMap(_.unboxIfPrimitive)
+        rhs <- evaluate(tree.rhs).flatMap(_.unboxIfPrimitive)
+        loader <- frame.classLoader()
+        result <- tree.op.evaluate(lhs, rhs, loader)
+      } yield result
+
+    @inline private def invokeMethod(tree: InstanceMethodTree): Safe[JdiValue] =
+      for {
+        qualValue <- evaluate(tree.qual)
+        argsValues <- tree.args.map(evaluate).traverse
+        argsBoxedIfNeeded <- JdiObject.boxUnboxOnNeed(tree.method.argumentTypes(), argsValues, frame)
+        result <- qualValue.asObject.invoke(tree.method, argsBoxedIfNeeded)
+      } yield result
+
+    /* -------------------------------------------------------------------------- */
+    /*                              Module evaluation                             */
+    /* -------------------------------------------------------------------------- */
+    @inline private def evaluateModule(tree: ModuleTree): Safe[JdiObject] =
+      Safe(JdiObject(tree.`type`.instances(1).get(0), frame.thread))
+
+    /* -------------------------------------------------------------------------- */
+    /*                                Instantiation                               */
+    /* -------------------------------------------------------------------------- */
+    @inline private def instantiate(tree: NewInstanceTree): Safe[JdiObject] =
+      for {
+        args <- tree.args.map(evaluate).traverse
+        boxedUnboxedArgs <- JdiObject.boxUnboxOnNeed(tree.method.argumentTypes(), args, frame)
+        instance <- JdiClass(tree.`type`, frame.thread).newInstance(tree.method, boxedUnboxedArgs)
+      } yield instance
+  }
+
+  /* -------------------------------------------------------------------------- */
+
+  private object RuntimeValidation {
+    case class Call(fun: Term, argClause: Term.ArgClause)
+    case class PreparedCall(qual: Validation[RuntimeValidationTree], name: String)
+
+    def validate(expression: Stat): Validation[RuntimeValidationTree] = {
+      expression match {
+        case value: Term.Name => validateName(value)
+        case _: Term.This => thisTree.map(tree => ThisTree(tree.value))
+        case sup: Term.Super => ???
+        case _: Term.Apply | _: Term.ApplyInfix | _: Term.ApplyUnary => validateMethod(extractCall(expression))
+        case select: Term.Select => validateSelect(select)
+        case newValue: Defn.Val => ???
```

Comment:
It should return some Recoverable instead of `???`

```suggestion
        case sup: Term.Super => ???
```

Reply:
I'll do them afterwards